### PR TITLE
ClamAV Cloud Function refactor [DDP-7621]

### PR DESCRIPTION
**Pull request by bskinner**
DDP-7621

This PR updates the ClamAV malware scanning Google Cloud Function to rely on socket communication with a independent instance running clamd, instead of packing the binaries inside the Cloud Function. There will be an follow-up pull request containing addition documentation and setup updates.

### Running ClamAV
In order to perform a test of this function, you will need to run a ClamAV instance locally.
* [Configure](https://cloud.google.com/artifact-registry/docs/docker/authentication) Docker to interact with the Google Artifact Registry
* Execute the below `docker run` command in a separate Terminal window. When testing is done, you may press `^C` to exit the container
  ```
  docker run \
      --interactive \
      --tty \
      --rm \
      --publish 3310:3310 \
      --name "clamav" \
      clamav/clamav:0.104
  ```

### Running the function
* Open a new terminal window and navigate to `ddp-study-server/cloud-functions/cf-file-scanner`
* From there, compile & run the function:
  ```
  GCP_PROJECT="broad-ddp-dev" \
    DDP_CLAMAV_SERVER="127.0.0.1:3310" \
    mvn -Drun.port=8080 function:run
  ```
* When testing is complete, you may press `^C` to exit the function.

### Invoking the function
In order to trigger a run of the function, a JSON event should be sent to the local running instance of the function. This can be done using `curl`:

#### Positive test
_(should positively identify `WIN.TEST.EICAR_HDB-1`)_
##### Shell command
```
curl -X POST 127.0.0.1:8080 \
  -H 'content-type: application/json' \
  -d '{
    "context": {
      "eventId": "11235813853211",
      "timestamp": "1999-12-31T11:59:99.999Z",
      "eventType": "google.pubsub.topic.publish",
      "resource": {
        "service": "pubsub.googleapis.com",
        "name": "projects/broad-ddp-dev/topics/cf-file-scanner",
        "type": "type.googleapis.com/google.pubsub.v1.PubsubMessage"
      }
    },
    "data": {
      "@type": "type.googleapis.com/google.pubsub.v1.PubsubMessage",
      "attributes": {
        "bucketId": "ddp-dev-file-uploads",
        "objectId": "test-files/positive-eicar.txt",
        "other": "attributes"
      },
      "data": ""
    }
  }'
```

##### Expected output
```
INFO: Received message: eventId=11235813853211 timestamp=1999-12-31T11:59:99.999Z
INFO: Looking up file: bucket=ddp-dev-file-uploads name=test-files/positive-eicar.txt
INFO: Found file that was created at 2022-03-10T20:15:42.576Z
INFO: Scanning file: BlobId{bucket=ddp-dev-file-uploads, name=test-files/positive-eicar.txt, generation=1646943342567482}
INFO: received response from clamd: 'stream: Win.Test.EICAR_HDB-1 FOUND'
SEVERE: malware variant identified Win.Test.EICAR_HDB-1 FOUND
SEVERE: message 11235813853211: malware was identified in object ddp-dev-file-uploads/test-files/positive-eicar.txt
INFO: PubSub topic not configured, not sending a scan result message
```

##### Shell command
```
curl -X POST 127.0.0.1:8080 \
  -H 'content-type: application/json' \
  -d '{
    "context": {
      "eventId": "1123581321342113853211",
      "timestamp": "1999-12-31T11:59:99.999Z",
      "eventType": "google.pubsub.topic.publish",
      "resource": {
        "service": "pubsub.googleapis.com",
        "name": "projects/broad-ddp-dev/topics/cf-file-scanner",
        "type": "type.googleapis.com/google.pubsub.v1.PubsubMessage"
      }
    },
    "data": {
      "@type": "type.googleapis.com/google.pubsub.v1.PubsubMessage",
      "attributes": {
        "bucketId": "ddp-dev-file-uploads",
        "objectId": "test-files/negative.txt",
        "other": "attributes"
      },
      "data": ""
    }
  }'
```

##### Expected output
```
INFO: Received message: eventId=1123581321342113853211 timestamp=1999-12-31T11:59:99.999Z
INFO: Looking up file: bucket=ddp-dev-file-uploads name=test-files/negative.txt
INFO: Found file that was created at 2022-03-10T20:16:31.045Z
INFO: Scanning file: BlobId{bucket=ddp-dev-file-uploads, name=test-files/negative.txt, generation=1646943391037227}
INFO: received response from clamd: 'stream: OK'
INFO: PubSub topic not configured, not sending a scan result message
```

### CODE REVIEW

File: cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    private InetSocketAddress clamdAddress;
+    private int socketTimeout = Defaults.SO_TIMEOUT;
+
+    public Client(String host) {
+        this(host, Defaults.PORT);
+    }
+    
+    public Client(String host, int port) {
+        assert host != null;
+        assert host.isEmpty() == false;
```
Comment (bskinner) [inline cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java]: Replace these with throwing checks, or just leave the verification to `InetSocketAddress`

File: cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+                    // be any data for us as a client to read until after the
+                    // terminating zero-size chunk is written.
+                    if (clamdInputStream.available() > 0) {
+                        throw new IOException("unexpected response from clamd");
+                    }
+
+                    read = inputStream.read(dataBuffer.array(), 0, chunkSize);
+                }
+
+                logger.fine("data chunks sent, preparting termination chunk");
```
Comment (bskinner) [inline cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java]: Typo:

```suggestion
                logger.fine("data chunks sent, preparing termination chunk");
```

File: cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+                try (var scanner = new Scanner(clamdInputStream, StandardCharsets.US_ASCII)) {
+                    scanner.useDelimiter(new String(delimiter.lineTerminator(), StandardCharsets.US_ASCII));
+                    response = scanner.next().trim();
+                } catch (NoSuchElementException nsee) {
+                    // This exception will be thrown by Scanner if a token delimiter has not been
+                    // read before the connection time or the underlying socket has been closed
+                    throw new IOException("unexpected response from clamd", nsee);
+                }
+
+                logger.info(String.format("received response from clamd: '%s'", response));
```
Comment (bskinner) [inline cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java]: Use the concatenation operator, not `String.format` here

File: cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+                */
+
+                final var sizeLimitError = "INSTREAM size limit exceeded";
+                final var streamHeader = "stream:";
+                final var negativeResponse = "OK";
+
+                if (response.startsWith(sizeLimitError)) {
+                    throw new IOException("clamav file size limit exceeded");
+                } else if (response.startsWith(streamHeader) == false) {
+                    var message = String.format("unrecognized response from clamav: %s", response);
```
Comment (bskinner) [inline cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java]: Use the concatenation operator, not `String.format` here

File: cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+                    throw new IOException(message);
+                }
+
+                response = response.replaceFirst(streamHeader, "").trim();
+                
+                if (response.startsWith(negativeResponse)) {
+                    return new ScanResult(MalwareResult.NEGATIVE);
+                } else {
+                    logger.fine(String.format("malware detected: %s", response));
```
Comment (bskinner) [inline cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java]: Use the concatenation operator, not `String.format` here

**Comment by ssettipalli** [comment]
@bskinner 
Can you please check/update  ddp-study-server/cloud-functions/cf-file-scanner/README.md to reflect updated file scan process

**Comment by bskinner** [comment]
> @bskinner Can you please check/update ddp-study-server/cloud-functions/cf-file-scanner/README.md to reflect updated file scan process

That was mentioned in the PR description- The intent of this PR is getting the code changes reviewed, and there will be documentation updates coming with the updated deployment process. If merging should be held off until that documentation is in place, could you please leave a "Request Changes" review with that comment so it's easier for me to track.

File: cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            switch (this) {
+                case NULL:
+                    return new byte[] { 0x00 };
+                case NEWLINE:
+                    return new byte[] { 0x0A };
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        }
+    }
```
Comment (bskinner) [inline cloud-functions/cf-file-scanner/src/main/java/org/broadinstitute/ddp/clamav/Client.java]: Pretty sure this is a Swift-ism working it's way in. I can propose an alternative using `getModeFlag()` and `getLineTerminator()`, and using a ctor to pass in the values (which would get rid of that awkward throw), but is this an idiomatic enum use in our codebase to begin with?

**Review by ssettipalli — APPROVED** [review]
looks good to me

**Pull request merged**